charts/actions-runner-controller/templates/webhook_configs.yaml

```diff
@@ -219,3 +237,18 @@ webhooks:
     resources:
     - runnerreplicasets
   sideEffects: None
+{{ if not (and .Values.admissionWebHooks.caBundle .Values.certManagerEnabled) }}
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ include "actions-runner-controller.servingCertName" . }}
```

Comment:
I realized that controller-runtime kindly watches and reloads the tls.crt and tls.key on update!

https://sourcegraph.com/github.com/kubernetes-sigs/controller-runtime/-/blob/pkg/webhook/server.go?L221#tab=references

Note though, regardless of whether we use cert-manager or this cert generation feature, helm-upgrade might result in some downtime due to K8s secret volume mount takes some time, a minutes or so, according to https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ `--sync-frequency `, to reflect changes made in the corresponding k8s secret.

It would be even greater if you could make the chart to regenerate the ca, the cert and the key only when only on expiration. Is that possible?

Reply:
Unfortunately the implementation of `genCA` does not cache the certificate between helm runs so we can be in either of two situations. We generate the CA in a helm hook so it is never updated by helm, this has the downside of having to also create the mutatingwebhookconfiguration as a helm hook due to helm scopes; or we can rotate the certificate each time causing a short outage to the web hook while ca rotation propagates.

The solution is intended to cover a specific corner case where you don't provide your own certs and you don't want to deploy cert-manager. My particular use-case is to have dedicated clusters for GitHub runners so an outage of the web hook wile we refresh the chart only causes minimal delays in spinning up workers, on clusters with high pod churn rates I can see this being a problem but if it is properly documented and called out I think it is a liveable tradeoff.

Reply:
Got it! Thanks a lot for your clarification. I do think it's a liveable tradeoff as long as it's well documented as you've already one in this PR 👍 I'm going to merge this.

BTW, what if we enhanced ARC to embed it's own cert manager (not jetstack/cert-manager) as I've outlined in https://github.com/actions-runner-controller/actions-runner-controller/pull/1780#discussion_r975920608 and enabled ARC to create/update the mutatingwebhookconfigs by itself? It would potentially simplify ARC deployments, in exchange of additional permissions required by ARC(CRUD permissions on cluster-wide mutatingwebhookconfig resources, which may be huge and not everyone could allow it).
Would you use it if we managed to implement it?

Reply:
We already use kyverno that has this approach, thus why I gave it as an example above so for our use-case the extra permissions would not be an issue but we could implement conditional logic in ARC so that it can create the mutatingwebhook configuration when permissions are granted and still rely on user provided certs with helm created web hook when the user does not want to grant those extra permissions. It would seem to complicate the logic though.